### DISCUSSION TITLE
[Q] Read date cell with format from file

### DISCUSSION THREAD

**Discussion by hdiaz-nectia** [exceljs/exceljs#2398] [Q&A]
## 💬 Questions and Help

I've a problem when I'm reading a Date value from columns in a excel file, I can't get the truth value in the correct format like a Date data type, instead I get a number.

My project is build with NestJS, my code is:

```js
worksheet.eachRow({ includeEmpty: true }, (row, rowNumber) => {
	row.eachCell((cell, colNumber) => {
		...
	    const header = headers[colNumber - 1];
            switch (header) {
               ...
               case 'dateColumn':
            	    // Date format is dd/mm/yyyy
            	    // Original value in excel is '16/06/2020'
	            const cellValue: Date = cell.value as Date; // When i get value from cell is 43998
	            console.log(cellValue); // print 43998
	            break;
	        ...
	}
}
```

I don't know how receive this  value like a string to manually convert to Date.

Any suggestion?

**Comment by nish17** [comment]
facing the same issue

any update @hdiaz-nectia ?

**Comment by nyuwec** [comment]
I have the same issue, if I use the non-stream reader i can get back a string date in the format of 'YYYY-MM-DDTHH:MM:SSZ' but a number through the stream reader. Also the cell type is 2 (number) and not 4 (date).
in the above example try:
```
console.log(cell.type)
```

**Comment by nyuwec** [comment]
actually there is a workaround, that float value can be parsed into a JS date.
You need to get the cell value as a number, so you'll get something like: `43392.7200967477`

```
new Date((excelDate - (25567 + 2))*86400000);
```

explanation:
Excel stores the dates in a float number, the decimal part is the elapsed days since 1900-01-01.
The remainder is the fragment of a day, so if you want to calculate it as millisecs you have to multiply this value with 86400000, if you need seconds use 86400.

25567 is the diff between JS epoch of 1970-01-01 and 1900-01-01

+2 is needed due to some leapyear bug in excel or in JS, i could not figure that out... yes, not scientifically proven, but close enough for most of the cases.
